keras_nlp/metrics/perplexity.py

```diff
@@ -0,0 +1,166 @@
+            computing the cross entropy loss). Note that if this field is
+            provided, the `sample_weight` field in `update_state()` is ignored.
+        **kwargs: Other keyword arguments.
+
+    Examples:
+
+    1. Calculate perplexity by calling update_state() and result().
+    1.1. `sample_weight`, and `pad_token_id` are not provided.
+    >>> tf.random.set_seed(42)
+    >>> perplexity = keras_nlp.metrics.Perplexity(name="perplexity")
+    >>> target = tf.random.uniform(
+    ...     shape=[2, 5],  maxval=10, dtype=tf.int32, seed=42)
+    >>> logits = tf.random.uniform(shape=(2, 5, 10), seed=42)
+    >>> perplexity.update_state(target, logits)
+    >>> perplexity.result()
+    <tf.Tensor: shape=(), dtype=float32, numpy=11.8781595>
+
+    1.2. `sample_weight` specified (masking token with ID 0).
+    >>> tf.random.set_seed(42)
+    >>> perplexity = keras_nlp.metrics.Perplexity(name="perplexity")
+    >>> target = tf.random.uniform(
+    ...     shape=[2, 5],  maxval=10, dtype=tf.int32, seed=42)
+    >>> logits = tf.random.uniform(shape=(2, 5, 10), seed=42)
+    >>> sample_weight = tf.cast(
+    ...     tf.math.logical_not(tf.equal(target, 0)), tf.float32)
+    >>> perplexity.update_state(target, logits, sample_weight)
+    >>> perplexity.result()
+    <tf.Tensor: shape=(), dtype=float32, numpy=13.1128>
+
+    2. Call perplexity directly.
+    >>> tf.random.set_seed(42)
+    >>> perplexity = keras_nlp.metrics.Perplexity(name="perplexity")
+    >>> target = tf.random.uniform(
+    ...     shape=[2, 5],  maxval=10, dtype=tf.int32, seed=42)
+    >>> logits = tf.random.uniform(shape=(2, 5, 10), seed=42)
+    >>> perplexity(target, logits)
+    <tf.Tensor: shape=(), dtype=float32, numpy=11.8781595>
+
+    3. Provide the padding token ID and let the class compute the mask on its
+       own.
+    >>> tf.random.set_seed(42)
+    >>> perplexity = keras_nlp.metrics.Perplexity(
+    ...     name="perplexity", pad_token_id=0)
+    >>> target = tf.random.uniform(
+    ...     shape=[2, 5],  maxval=10, dtype=tf.int32, seed=42)
+    >>> logits = tf.random.uniform(shape=(2, 5, 10), seed=42)
+    >>> perplexity(target, logits)
+    <tf.Tensor: shape=(), dtype=float32, numpy=13.1128>
+    """
+
+    def __init__(
+        self,
+        name="perplexity",
+        dtype=None,
+        from_logits=False,
+        pad_token_id=None,
+        **kwargs,
+    ):
+        super().__init__(name=name, dtype=dtype, **kwargs)
+
+        if not tf.as_dtype(self.dtype).is_floating:
+            raise ValueError(
+                "`dtype` must be a floating point type. "
+                f"Received: dtype={dtype}"
+            )
+
+        self._cross_entropy = keras.losses.SparseCategoricalCrossentropy(
+            from_logits=from_logits, reduction="sum"
+        )
+
+        self.pad_token_id = pad_token_id
+
+        self._aggregate_cross_entropy = self.add_weight(
+            name="aggregate_cross_entropy",
```

Comment:
Nit: Spell "crossentropy" in a single word, for consistency. This applies to the weight name and also to Python variable names.